subworkflows/local/eggnog.nf

```diff
@@ -0,0 +1,35 @@
+//
+// Run eggnog-mapper on called ORFs, first optionally downloading the required databases
+//
+
+include { EGGNOG_DOWNLOAD } from '../../modules/local/eggnog/download'
+include { EGGNOG_MAPPER   } from '../../modules/local/eggnog/mapper'
+include { EGGNOG_SUM      } from '../../modules/local/eggnog/sum'
+
+workflow EGGNOG {
+    take:
+    faa
+    collect_fcs
+    dbpath
+
+    main:
+    ch_versions = Channel.empty()
+
+    ch_dbpath = Channel.fromPath(dbpath, checkIfExists: true)
+
+    EGGNOG_DOWNLOAD()
```

Comment:
Hi @mahesh-panchal  we made changes in order to avoid the if clause and getting the download through storDir.
take a look!

Reply:
I'm confused. I thought you wanted either the user path, or the download to happen.
If the user provides a path in this case and there isn't something in the EGGNOG_DOWNLOAD storedir path, then it will start the EGGNOG_DOWNLOAD process.

Reply:
when `eggnog_dbpath` contains the database, it will skip `EGGNOG_DOWNLOAD` because of `storeDir` while if the folder is empty, `EGGNOG_DOWNLOAD` will download the database.

Reply:
Ah, ok. I understand now. You make the db parameter a required input, and use `storeDir` to trigger `EGGNOG_DOWNLOAD`.  Do you need the `ch_dbpath` then as input to EGGNOG_MAPPER? If the user provides a database, and `storeDir` finds it, then isn't the path in `EGGNOG_DOWNLOAD.out.eggnog_db` anyway?

Reply:
`EGGNOG_DOWNLOAD.out.eggnog_db' is used as marker but it won't be used by `EGGNOG_MAPPER`

Reply:
`EGGNOG_DOWNLOAD.out.eggnog_db` is just an output file from `EGGNOG_DOWNLOAD` but it isn't the folder used by `EGGNOG_MAPPER`. When EGGNOG database is present in the folder, `EGGNOG_DOWNLOAD` will emit the file already present in EGGNOG folder skipping the module process. By having both values we ensure that `EGGNOG_MAPPER` runs only when the database is downloaded or stored.

Reply:
What Danilo is saying is that we need a way to make sure `EGGNOG_MAPPER` waits for `EGGNOG_DOWNLOAD` as the download directory is not emitted by the `EGGNOG_DOWNLOAD` module any more.

The reason `EGGNOG_DOWNLOAD` for that is that we need to download directly to the work directory to avoid getting a subdirectory in the target directory and it doesn't seem one can emit the work directory. (We tried emitting `'.'` but that didn't work. I suppose Nextflow doesn't want to symlink a full work directory into another process' work.)

An alternative to adding a "marker" variable to the `EGGNOG_DOWNLOAD` signature would be to do a bit of channel algebra combining the `EGGNOG_DOWNLOAD.out.eggnog_db` with the `ch_dbpath` channel but end up passing just the latter. The advantage of this is of course that the signature of the `EGGNOG_MAPPER` process would be cleaner; the disadvantage would be the unnecessary channel algebra.

I hope this makes it clearer.

Reply:
You should just be able to specify the files and Nextflow will check directory contents for you. You shouldn't need to emit the current directory as well
See this toy:
```nextflow
workflow {
    EGGNOG_DOWNLOAD 
    | EGGNOG_MAPPER
    | view
}

process EGGNOG_DOWNLOAD {
    input:

    script:
    """
    touch eggnog{.db,_proteins.dmnd,.taxa.db,.taxa.db.traverse.pkl}
    """

    output:
    path("eggnog{.db,_proteins.dmnd,.taxa.db,.taxa.db.traverse.pkl}", arity: '4')
}

process EGGNOG_MAPPER {
    input:
    path eggnog_files, stageAs: 'eggnog/*'

    script:
    """
    ls -R eggnog
    """

    output:
    stdout
}
```
```
params.eggnogdb = 'databases/eggnog'

process {
    withName: 'EGGNOG_DOWNLOAD' {
        storeDir = "${params.eggnogdb}"
    }
}
```
If the eggnog folder exists, but is empty Nextflow runs the download. If Nextflow finds some of those files but not all, it throws an error. If all the files are there, it runs the download. In all cases EGGNOG_MAPPER has to wait for EGGNOG_DOWNLOAD, and there's no need for an extra path to be supplied to EGGNOG_MAPPER.

I don't think there's any need for the channel algebra.

Reply:
@mahesh-panchal we updated the subworkflow based on your comments.

Reply:
Given this is a local module, wouldn't it be simpler to output all the files in a single channel, and leave out the channel algebra? E.g., use your `all` channel but specify `arity:'4'` in the type declaration. Sorry, I don't want to drag this out.

Reply:
No problem! we discussed about using `all` while making this change and we preferred to explicit the output.
What is the advantage of changing it in the way you suggest?

Reply:
There's little difference other than having a little less to maintain. However, if you'd prefer this, then let's stick with that.

Reply:
than we keep it as it is now!